src/engine/strat_engine/liminal.rs

```diff
@@ -422,13 +761,208 @@ impl LiminalDevices {
+                            (LInfo::Stratis(strat_info_1), LInfo::Stratis(strat_info_2));
+                        let err_msg = format!(
+                            "Information about previously discovered device {} incompatible with information about newly discovered device {}",
+                            info_1,
+                            info_2);
+                        Err((err_msg, info_1, info_2))
+                    } else {
+                        Ok(LInfo::Stratis(LStratisInfo {
+                            ids: strat_info_2.ids,
+                            luks: strat_info_2.luks.or(strat_info_1.luks),
+                        }))
+                    }
+                }
+            }
+        }
+
+        let stratis_identifiers = info.stratis_identifiers();
+        let device_uuid = stratis_identifiers.device_uuid;
+
+        match devices.remove(&device_uuid) {
+            None => {
+                info!(
+                    "device with Stratis identifiers {} discovered, i.e., identified for the first time during this execution of stratisd",
+                    stratis_identifiers);
+                devices.insert(device_uuid, info);
+                Ok(devices)
+            }
+            Some(removed) => match combine_two_devices(removed, info) {
+                Err((err, removed, info)) => {
+                    warn!(
+                        "Moving set of devices with pool UUID {} to hopeless sets because {}",
+                        stratis_identifiers.pool_uuid, err
+                    );
+                    let mut hopeless: HashSet<LInfo> =
+                        devices.drain().map(|(_, info)| info).collect();
+                    hopeless.insert(removed);
+                    hopeless.insert(info);
+                    Err(hopeless)
+                }
+                Ok(info) => {
+                    devices.insert(device_uuid, info);
+                    Ok(devices)
+                }
+            },
+        }
+    }
+
+    /// Process a device for removal from a set of devices.
+    fn process_info_remove(&mut self, devices: &mut HashMap<DevUuid, LInfo>, info: LInfo) {
+        // Combine two devices which have identical pool and device UUIDs.
+        // The first argument is the existing information, the second is the
+        // information about the removed device, where "removed" means there
+        // was a udev "remove" event and this info has been found out about the
+        // device attached to the event.
+        fn combine_remove_devices(info_1: LInfo, info_2: LInfo) -> Option<LInfo> {
+            match (info_1, info_2) {
+                (LInfo::Luks(luks_info), LInfo::Stratis(_)) => Some(LInfo::Luks(luks_info)),
+                (LInfo::Stratis(strat_info), LInfo::Luks(_)) => {
+                    Some(LInfo::Stratis(LStratisInfo {
+                        ids: strat_info.ids,
+                        luks: None,
+                    }))
+                }
```

Comment:
I'm not sure what to make of this handling mainly because it seems like this would only occur if a closed encrypted device has the same identifiers as a physical device that's been registered with Stratis. Can you explain what the motivation was for this?

Reply:
> I'm not sure what to make of this handling mainly because it seems like this would only occur if a closed encrypted device has the same identifiers as a physical device that's been registered with Stratis. Can you explain what the motivation was for this?

It doesn't have any good physical explanation. I'll plug in some warnings, and we can see if they ever turn up.